src/test/java/com/github/checkstyle/regression/git/DiffParserTest.java

```diff
@@ -81,6 +98,50 @@ public void testParseDeleteChange() throws Exception {
     }
 
     @Test
+    public void testParseRenameChange() throws Exception {
+        try (Repository repository = GitUtils.createNewRepository()) {
+            GitUtils.addAnEmptyFileAndCommit(repository, "HelloWorld");
+            GitUtils.createNewBranchAndCheckout(repository, "foo");
+            GitUtils.addAnEmptyFileAndCommit(repository, "HelloWorldFoo");
+            GitUtils.removeFileAndCommit(repository, "HelloWorld");
+            final List<GitChange> changes = DiffParser.parse(
+                    repository.getDirectory().getParent(), "foo");
+            assertEquals("There should be 1 change detected", 1, changes.size());
+            final GitChange expected = ImmutableGitChange.builder()
+                    .path("HelloWorldFoo")
+                    .build();
+            assertEquals("The change is not as expected", expected, changes.get(0));
+        }
+    }
+
+    @Test
+    public void testParseCopyChange() throws Exception {
+        try (Repository repository = GitUtils.createNewRepository()) {
+            GitUtils.addAnEmptyFileAndCommit(repository, "a.txt");
+            GitUtils.createNewBranchAndCheckout(repository, "foo");
+            GitUtils.addAnEmptyFileAndCommit(repository, "b.txt");
+            GitUtils.addAnEmptyFileAndCommit(repository, "src/com/foo/c.java");
+            GitUtils.addAnEmptyFileAndCommit(repository, "src/com/foo/d.java");
+            GitUtils.removeFileAndCommit(repository, "a.txt");
```

Comment:
I was puzzled when I first investigating on `COPY`. It has the following rules:
- If no files deleted, then anything new would be detected as `ADD`.
- If one file is deleted, then one file added, and they are the "same"(in Git context), then it is a `RENAME`.
- If one file is deleted, some files added, and they the "same", then one of them is a `RENAME`, others is `COPY`.

> how can we tell which one is being copied where?

I don't exactly know, since the "same" in Git is describing as `score`. I guess the `RENAME` is one with the highest score and been added to staging area earliest.

Actually this UT is almost a copy of https://github.com/eclipse/jgit/blob/005e5feb4ecd08c4e4d141a38b9e7942accb3212/org.eclipse.jgit.test/tst/org/eclipse/jgit/diff/RenameDetectorTest.java#L209